### DISCUSSION TITLE
Binary data issue?

### DISCUSSION THREAD

**Discussion by hancockks** [xlate/staedi#389] [General]
Not sure if this should be an issue, but having difficulty finding X12 examples with binary data. I stumbled across this one on stack overflow and after making sure spaces->tabs, it looks like it's valid but staedi is throwing an odd error.  Walking the code it doesn't look like it trying to parse the binary?

EDIE003 - Invalid processing state; INVALID (previous: TAG_SEARCH); input: '<' in segment BIN at position 33

https://stackoverflow.com/questions/807871/does-anyone-have-a-sample-file-in-x12-or-edifact-that-contains-binary-data

https://gist.github.com/hancockks/9b8c7199bf0d67860a8f8a5c7488fab1

**Comment by MikeEdgar** [comment]
Hi @hancockks , are you either using a schema that identifies `BIN01` as the binary data length or calling `setBinaryDataLength` with the value of `BIN01` in your code?

**Reply by hancockks** [reply]
Neither -- I'm guessing from your question I'm not handling it correctly. I assumed I'd get a event with EVENT_DATA_BINARY at which point I'd call reader.getBinaryData().readAllBytes() to grab the data:

case ELEMENT_DATA_BINARY:
  currentSegment.addData(reader.getBinaryData().readAllBytes());
  break;

I'm not injecting any schemas at the transaction level.

**Reply by MikeEdgar** [reply]
That code should work, but you'll need to do something like this first in your read loop (adding any validation you'd like as needed).

```java
case ELEMENT_DATA:
    if ("BIN".equals(reader.getLocation().getSegmentTag() && reader.getLocation().getElementPosition() == 1) {
        long binaryDataLength = Long.parseLong(reader.getText());
        reader.setBinaryDataLength(binaryDataLength);
    }
    break;
```

**Reply by hancockks** [reply]
Thanks! That seems to do the trick.

I incorrectly assumed that I'd get either a ELEMENT_DATA OR a ELEMENT_DATA_BINARY, not both.

**Reply by MikeEdgar** [reply]
Right, the parser needs some extra information to know when the thing it's about to parse is actually binary data. It's possible with schemas or the way we discussed :+1:
